FR-6132 - Fix resume feature

upils opened the pull request:
Fixing https://bugs.launchpad.net/ubuntu-image/+bug/2043482

Summary of changes (not obvious since I moved some functions around):

- `calculateStates` and `parseImageDefinition` are not state functions anymore.  They are "setup" functions. I find that stateFunctions should map to states and so map to a specific state of the resulting image. This was not the case for both of these functions. This helps avoid complexity to deal with the resume feature since we can properly calculate the full list of states and then resume where we stopped. This should also be clearer to our users that the execution of a state is relfected on the resulting image. 
- Packages and Snaps installed/to install are not properly saved
- The tracking of the CurrentStep never worked before and was not used. Now this is tracked properly, but only used to report it when saving the state.

sil2100 reviewed (dismissed):
At first I wasn't sure about the move of `parseImageDefinition` and `calculateStates`, but I guess this way makes more sense. Sure, we're adding more to the state machine preparation, but conceptually it was a bit weird that states that dynamically created the state machine were also states. I think the usage of breaking at those states is really unlikely.
Also, wonder how we didn't catch the `VolumeNames` missing in the load state method. Was it that we only checked for its existence in the saved state and not in the loaded one?

upils commented:
> At first I wasn't sure about the move of `parseImageDefinition` and `calculateStates`, but I guess this way makes more sense. Sure, we're adding more to the state machine preparation, but conceptually it was a bit weird that states that dynamically created the state machine were also states. I think the usage of breaking at those states is really unlikely.

I agree and I had the same reasoning. Moreover from a user point of view, I find it clearer.

> Also, wonder how we didn't catch the VolumeNames missing in the load state method. Was it that we only checked for its existence in the saved state and not in the loaded one?

We didn't catch it because I missed it when I reworked the read/write metadata feature. The value was nil in the JSON loaded to test. I checked the other fields and added Packages/Snaps that were missing. The others missing are missing on purpose (loaded/updated somewhere else). Ideally we should find a way to detect that we test every field with a non-nil value to make sure a test break if in the future we had a new field and we forget to load it from metadata. 

I suspect only a few users use it so it went under the radar all this time.
